lib/json/pure/generator.rb

```diff
@@ -286,12 +288,70 @@ def to_h
         # created this method raises a
         # GeneratorError exception.
         def generate(obj)
-          result = obj.to_json(self)
+          if @indent.empty? and @space.empty? and @space_before.empty? and @object_nl.empty? and @array_nl.empty? and
+              !@ascii_only and !@script_safe and @max_nesting == 0 and !@strict
+            result = generate_json(obj, '')
+          else
+            result = obj.to_json(self)
+          end
           JSON.valid_utf8?(result) or raise GeneratorError,
             "source sequence #{result.inspect} is illegal/malformed utf-8"
           result
         end
 
+        # Handles @allow_nan, @buffer_initial_length, other ivars must be the default value (see above)
+        private def generate_json(obj, buf)
+          case obj
+          when Hash
+            buf << '{'.freeze
+            first = true
+            obj.each_pair do |k,v|
+              buf << ','.freeze unless first
+              fast_serialize_string(k.to_s, buf)
+              buf << ':'.freeze
+              generate_json(v, buf)
+              first = false
+            end
+            buf << '}'.freeze
+          when Array
+            buf << '['.freeze
+            first = true
+            obj.each do |e|
+              buf << ','.freeze unless first
+              generate_json(e, buf)
+              first = false
+            end
+            buf << ']'.freeze
+          when String
+            fast_serialize_string(obj, buf)
+          when Integer
+            buf << obj.to_s
+          else
+            # Note: Float is handled this way since it is Float#to_s is slow anyway
+            buf << obj.to_json(self)
+          end
+        end
+
+        # Assumes !@ascii_only, !@script_safe
+        if Regexp.method_defined?(:match?)
+          private def fast_serialize_string(string, buf) # :nodoc:
+            buf << '"'.freeze
+            string = string.encode(::Encoding::UTF_8) unless string.encoding == ::Encoding::UTF_8
+
+            if /["\\\x0-\x1f]/n.match?(string)
+              buf << string.gsub(/["\\\x0-\x1f]/n, MAP)
+            else
+              buf << string
+            end
+            buf << '"'.freeze
+          end
+        else
+          # Ruby 2.3 compatibility
```

Comment:
At this point we might consider bumping the required version to 2.4 or even 2.5

Reply:
That would be nice, but this is not much overhead, so that could be cleaned up later when we bump the min Ruby version.